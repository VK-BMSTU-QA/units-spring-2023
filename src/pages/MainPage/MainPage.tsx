import React, { useState } from 'react';

import type { Category, SortBy } from '../../types';

import { useCurrentTime, useProducts } from '../../hooks';
import { Categories, ProductCard, SortButton } from '../../components';
import {
    applyCategories,
    updateCategories,
    getNextSortBy,
    productComparator, 
} from '../../utils';

import './MainPage.css';

export const MainPage = () => {
    const products = useProducts();
    const currentTime = useCurrentTime();

    const [selectedCategories, setSelectedCategories] = useState<Category[]>(
        []
    );
    const [sortBy, setSortBy] = useState<SortBy>('по умолчанию');

    const onCategoryClick = (clickedCategory: Category) =>
        setSelectedCategories(
            updateCategories(selectedCategories, clickedCategory)
        );

    const onSortButtonClick = () => setSortBy(getNextSortBy(sortBy));

    products.sort(productComparator(sortBy));

    return (
        <div className="main-page">
            <h2 className="main-page__title">VK Маркет</h2>
            <h3>{currentTime}</h3>
            <div className="main-page__parameters">
                <Categories
                    selectedCategories={selectedCategories}
                    onCategoryClick={onCategoryClick}
                />
                <SortButton
                    currentSort={sortBy}
                    onSortButtonClick={onSortButtonClick}
                />
            </div>
            {applyCategories(products, selectedCategories).map((product) => (
                <ProductCard key={product.id} {...product} />
            ))}
        </div>
    );
};
