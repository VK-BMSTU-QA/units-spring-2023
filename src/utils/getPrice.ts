import type { PriceSymbol } from '../types';

export const getPrice = (value: number, symbol: PriceSymbol = 'â‚½'): string =>
    `${value.toLocaleString('en-En')} ${symbol}`;
