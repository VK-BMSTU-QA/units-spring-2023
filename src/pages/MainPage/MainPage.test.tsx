import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { applyCategories, getNextSortBy } from '../../utils';
import { MainPage } from './MainPage';
import { Category, PriceSymbol, Product } from '../../types';

const products: Product[] = [
    {
        id: 1,
        name: 'some product 1',
        description: 'some description',
        price: 25000,
        priceSymbol: '$',
        category: 'Для дома',
    },
    {
        id: 2,
        name: 'some product 2',
        description: 'some description',
        price: 5000,
        priceSymbol: '$',
        category: 'Для дома',
        imgUrl: './someImg.png',
    },
    {
        id: 3,
        name: 'some product 3',
        description: 'some description',
        price: 200000,
        priceSymbol: '₽',
        category: 'Электроника',
    },
];

afterEach(jest.clearAllMocks);
afterEach(jest.useRealTimers);

jest.useFakeTimers().setSystemTime(new Date(2023, 3, 12, 12, 0, 0, 0));
jest.mock('../../hooks/useProducts', () => {
    return {
        __esModule: true,
        useProducts: jest.fn(() => products),
    };
});
jest.mock('../../utils/getPrice', () => {
    return {
        __esModule: true,
        getPrice: jest.fn(
            (value: number, symbol: PriceSymbol = '₽'): string =>
                `${value.toLocaleString('ru-RU')} ${symbol}`
        ),
    };
});
jest.mock('../../utils/getNextSortBy', () => {
    return {
        __esModule: true,
        getNextSortBy: jest.fn(() => 'по возрастанию цены'),
    };
});
jest.mock('../../utils/applyCategories', () => {
    return {
        __esModule: true,
        applyCategories: jest.fn(
            (products: Product[], categories: Category[]): Product[] => {
                if (!categories.length) return products;

                return products.filter((product) =>
                    categories.includes(product.category)
                );
            }
        ),
    };
});
describe('test mainPage', () => {
    // eslint-disable-next-line react/react-in-jsx-scope
    const rendered = render(<MainPage />);

    it('rendering', () => {
        expect(rendered.asFragment()).toMatchSnapshot();
    });

    it('getNextSortBy', () => {
        // eslint-disable-next-line react/react-in-jsx-scope
        const sortRendered = render(<MainPage />);

        expect(getNextSortBy).toHaveBeenCalledTimes(0);
        fireEvent.click(sortRendered.getByText('Сортировать по умолчанию'));
        expect(getNextSortBy).toHaveBeenCalledTimes(1);
    });
});
