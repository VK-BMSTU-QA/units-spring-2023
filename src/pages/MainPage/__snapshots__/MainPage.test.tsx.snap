// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test mainPage rendering 1`] = `
<DocumentFragment>
  <div
    class="main-page"
  >
    <h2
      class="main-page__title"
    >
      VK Маркет
    </h2>
    <h3>
      12:00:00
    </h3>
    <div
      class="main-page__parameters"
    >
      <div
        class="categories"
      >
        <div
          class="categories__badge"
        >
          Одежда
        </div>
        <div
          class="categories__badge"
        >
          Для дома
        </div>
        <div
          class="categories__badge"
        >
          Электроника
        </div>
      </div>
      <button
        class="sort-button"
      >
        Сортировать по умолчанию
      </button>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          some product 1
        </h2>
        <div
          class="product-card__description"
        >
          some description
        </div>
        <h1
          class="product-card__price"
        >
          25 000 $
        </h1>
        <div
          class="product-card__category"
        >
          Для дома
        </div>
      </div>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          some product 2
        </h2>
        <div
          class="product-card__description"
        >
          some description
        </div>
        <h1
          class="product-card__price"
        >
          5 000 $
        </h1>
        <div
          class="product-card__category"
        >
          Для дома
        </div>
      </div>
      <img
        alt="some product 2"
        class="product-card__image"
        src="./someImg.png"
      />
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          some product 3
        </h2>
        <div
          class="product-card__description"
        >
          some description
        </div>
        <h1
          class="product-card__price"
        >
          200 000 ₽
        </h1>
        <div
          class="product-card__category"
        >
          Электроника
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
